# Data Cleaning 

#### 1. Import pandas library.

In [2]:
import pandas as pd


In [3]:
pwd

'/Users/ayubpathan/Desktop/ironhack/Week2/lab-data-cleaning/your-code'

#### 2. Import the users table.

In [35]:
user_data = pd.read_csv('/Users/ayubpathan/Desktop/ironhack/Week2/lab-data-cleaning/data/users.csv')
user_data_original = user_data.copy()
user_data

Unnamed: 0     Id  Reputation  Views  UpVotes  DownVotes
0               0     -1           1      0     5007       1920
1               1      2         101     25        3          0
2               2      3         101     22       19          0
3               3      4         101     11        0          0
4               4      5        6792   1145      662          5
...           ...    ...         ...    ...      ...        ...
40320       40320  55743           1      0        0          0
40321       40321  55744           6      1        0          0
40322       40322  55745         101      0        0          0
40323       40323  55746         106      1        0          0
40324       40324  55747           1      0        0          0

[40325 rows x 6 columns]

#### 3. Rename Id column to userId.

In [29]:
list(user_data.columns)

['Unnamed: 0', 'Id', 'Reputation', 'Views', 'UpVotes', 'DownVotes']

In [45]:
user_data.rename(columns ={'Id':'userId'} , inplace = True)
list(user_data.columns)

['Unnamed: 0', 'userId', 'Reputation', 'Views', 'UpVotes', 'DownVotes']

In [46]:
user_data

Unnamed: 0  userId  Reputation  Views  UpVotes  DownVotes
0               0      -1           1      0     5007       1920
1               1       2         101     25        3          0
2               2       3         101     22       19          0
3               3       4         101     11        0          0
4               4       5        6792   1145      662          5
...           ...     ...         ...    ...      ...        ...
40320       40320   55743           1      0        0          0
40321       40321   55744           6      1        0          0
40322       40322   55745         101      0        0          0
40323       40323   55746         106      1        0          0
40324       40324   55747           1      0        0          0

[40325 rows x 6 columns]

#### 4. Import the posts table. 

In [36]:
post_data = pd.read_csv('/Users/ayubpathan/Desktop/ironhack/Week2/lab-data-cleaning/data/posts.csv')
post_data_original =post_data.copy()
post_data

Unnamed: 0      Id  OwnerUserId  Score  ViewCount  CommentCount
0               0       1          8.0     23     1278.0             1
1               1       2         24.0     22     8198.0             1
2               2       3         18.0     54     3613.0             4
3               3       4         23.0     13     5224.0             2
4               4       5         23.0     81        NaN             3
...           ...     ...          ...    ...        ...           ...
91971       91971  115374        805.0      2        NaN             2
91972       91972  115375      49365.0      0        9.0             0
91973       91973  115376      55746.0      1        5.0             2
91974       91974  115377        805.0      0        NaN             0
91975       91975  115378       7250.0      0        NaN             0

[91976 rows x 6 columns]

In [38]:
list(post_data.columns)


['Unnamed: 0', 'Id', 'OwnerUserId', 'Score', 'ViewCount', 'CommentCount']

In [48]:
post_data

Unnamed: 0  postId   userId  Score  ViewCount  CommentCount
0               0       1      8.0     23     1278.0             1
1               1       2     24.0     22     8198.0             1
2               2       3     18.0     54     3613.0             4
3               3       4     23.0     13     5224.0             2
4               4       5     23.0     81        NaN             3
...           ...     ...      ...    ...        ...           ...
91971       91971  115374    805.0      2        NaN             2
91972       91972  115375  49365.0      0        9.0             0
91973       91973  115376  55746.0      1        5.0             2
91974       91974  115377    805.0      0        NaN             0
91975       91975  115378   7250.0      0        NaN             0

[91976 rows x 6 columns]

#### 5. Rename Id column to postId and OwnerUserId to userId.

In [41]:
post_data.rename(columns={"Id":"postId", "OwnerUserId":"userId"}, inplace = True)
list(post_data.columns)

['Unnamed: 0', 'postId', 'userId', 'Score', 'ViewCount', 'CommentCount']

#### 6. Define new dataframes for users and posts with the following selected columns:
**users_sliced columns**: userId, Reputation, Views, UpVotes      
**posts_sliced columns**: postId, Score, userId, ViewCount

In [57]:
users_sliced_columns = user_data[["userId", "Reputation", "Views", "UpVotes"]]
users_sliced_columns

userId  Reputation  Views  UpVotes
0          -1           1      0     5007
1           2         101     25        3
2           3         101     22       19
3           4         101     11        0
4           5        6792   1145      662
...       ...         ...    ...      ...
40320   55743           1      0        0
40321   55744           6      1        0
40322   55745         101      0        0
40323   55746         106      1        0
40324   55747           1      0        0

[40325 rows x 4 columns]

In [80]:
posts_sliced_columns = post_data[["postId", "Score", "userId", "ViewCount"]]
posts_sliced_columns

postId  Score   userId  ViewCount
0           1     23      8.0     1278.0
1           2     22     24.0     8198.0
2           3     54     18.0     3613.0
3           4     13     23.0     5224.0
4           5     81     23.0        NaN
...       ...    ...      ...        ...
91971  115374      2    805.0        NaN
91972  115375      0  49365.0        9.0
91973  115376      1  55746.0        5.0
91974  115377      0    805.0        NaN
91975  115378      0   7250.0        NaN

[91976 rows x 4 columns]

#### 7. Merge the two dataframes created in the step above (8), users_sliced and posts_sliced. 
You will need to make a [merge](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.merge.html) of posts and users dataframes.

In [86]:
#merge_df = pd.DataFrame.merge(on='userId', posts_sliced_columns, users_sliced_columns, join ='Left').reset_index(drop=True)
merge_df = pd.DataFrame.merge(posts_sliced_columns, users_sliced_columns,how='inner',on='userId')
merge_df

postId  Score   userId  ViewCount  Reputation  Views  UpVotes
0           1     23      8.0     1278.0        6764   1089      604
1          16     16      8.0        NaN        6764   1089      604
2          36     41      8.0    67396.0        6764   1089      604
3          65     14      8.0        NaN        6764   1089      604
4          78     33      8.0        NaN        6764   1089      604
...       ...    ...      ...        ...         ...    ...      ...
90579  115366      1  55742.0       17.0           6      0        0
90580  115370      1  55744.0       13.0           6      1        0
90581  115371      0  35801.0       19.0           1      1        0
90582  115375      0  49365.0        9.0           1      0        0
90583  115376      1  55746.0        5.0         106      1        0

[90584 rows x 7 columns]

#### 8. How many missing values do you have in your merged dataframe? On which columns?

In [90]:
merge_df.isna().sum()

postId            0
Score             0
userId            0
ViewCount     48396
Reputation        0
Views             0
UpVotes           0
dtype: int64

### Total na values are 48396 on ViewCount column.

#### 9. You will need to make something with missing values.  Will you clean or filling them? Explain. 
**Remember** to check the results of your code before going to the next step.

#### It depends on use of data and what one has to do with the half of the data from same column where there  almost half of the values are missing. It is very difficult to predict to drop this whole column or simply fill all na with other values  Hence answer lies in dependecies. 
